src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionProvider.cs

```diff
@@ -0,0 +1,229 @@
+
+            // This hashset will be used to match namespace names, so it must have the same case-sensitivity as the source language.
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var namespacesInScope = new HashSet<string>(importedNamespaces, syntaxFacts.StringComparer);
+
+            // Get containing namespaces.
+            var namespaceSymbol = semanticModel.GetEnclosingNamespace(syntaxContext.Position, cancellationToken);
+            while (namespaceSymbol != null)
+            {
+                namespacesInScope.Add(namespaceSymbol.ToDisplayString(SymbolDisplayFormats.NameFormat));
+                namespaceSymbol = namespaceSymbol.ContainingNamespace;
+            }
+
+            return namespacesInScope;
+        }
+
+        internal override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem completionItem, TextSpan completionListSpan, char? commitKey, CancellationToken cancellationToken)
+        {
+            var containingNamespace = ImportCompletionItem.GetContainingNamespace(completionItem);
+
+            if (await ShouldCompleteWithFullyQualifyTypeName().ConfigureAwait(false))
+            {
+                var fullyQualifiedName = $"{containingNamespace}.{completionItem.DisplayText}";
+                var change = new TextChange(completionListSpan, fullyQualifiedName);
+
+                return CompletionChange.Create(change);
+            }
+            else
+            {
+                // Find context node so we can use it to decide where to insert using/imports.
+                var tree = (await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false))!;
+                var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+                var addImportContextNode = root.FindToken(completionListSpan.Start, findInsideTrivia: true).Parent;
+
+                // Add required using/imports directive.                              
+                var addImportService = document.GetLanguageService<IAddImportsService>();
+                var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
+                var compilation = (await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false))!;
+                var importNode = CreateImport(document, containingNamespace);
+
+                var rootWithImport = addImportService.AddImport(compilation, root, addImportContextNode, importNode, placeSystemNamespaceFirst, cancellationToken);
+                var documentWithImport = document.WithSyntaxRoot(rootWithImport);
+                var formattedDocumentWithImport = await Formatter.FormatAsync(documentWithImport, Formatter.Annotation, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
Probably worth mentioning that you annotated the import with this annotation, so this is only to format that bit of the document, and nothin else.